src/autopas/utils/SimilarityFunctions.h

```diff
@@ -28,86 +46,68 @@ template <class Container>
+  const auto numberOfBins = numberOfBinsPerDim * numberOfBinsPerDim * numberOfBinsPerDim;
+  const auto binVolume = domainVolume / (double)numberOfBins;
+  const auto binDimensions = domainDimensions / (double)numberOfBinsPerDim;
+
+  // Calculate the actual number of bins per dimension. The rounding is needed in case the division slightly
+  // underestimates the division. The choice of dimension 0 is arbitrary.
+  const int binsPerDimension = static_cast<int>(std::round(domainDimensions[0] / binDimensions[0]));
 
-  std::vector<size_t> particlesPerCell(numberOfCells, 0);
-  std::vector<double> allVolumes(numberOfCells, 0);
+  std::vector<size_t> particlesPerBin(numberOfBins, 0);
 
-  // add particles accordingly to their cell to get the amount of particles in each cell
+  // add particles accordingly to their bin to get the amount of particles in each bin
   for (auto particleItr = container.begin(autopas::IteratorBehavior::owned); particleItr.isValid(); ++particleItr) {
     const auto &particleLocation = particleItr->getR();
-    std::array<size_t, 3> index = {};
-    for (size_t i = 0; i < particleLocation.size(); i++) {
-      index[i] = particleLocation[i] / cellLength;
-    }
-    const size_t cellIndex = autopas::utils::ThreeDimensionalMapping::threeToOneD(index, cellsPerDimension);
-    particlesPerCell[cellIndex] += 1;
-    // calculate the size of the current cell
-    allVolumes[cellIndex] = 1;
-    for (size_t i = 0; i < cellsPerDimension.size(); ++i) {
-      // the last cell layer has a special size
-      if (index[i] == cellsPerDimension[i] - 1) {
-        allVolumes[cellIndex] *= outerCellSizePerDimension[i];
-      } else {
-        allVolumes[cellIndex] *= cellLength;
-      }
-    }
+
+    const auto binIndex3d = autopas::utils::ArrayMath::floorToInt((particleLocation - startCorner) / binDimensions);
+    const auto binIndex1d = autopas::utils::ThreeDimensionalMapping::threeToOneD(
+        binIndex3d, {binsPerDimension, binsPerDimension, binsPerDimension});
+
+    particlesPerBin[binIndex1d] += 1;
   }
 
-  // calculate density for each cell
+  // calculate density for each bin and track max density
   double maxDensity{0.};
-  std::vector<double> densityPerCell(numberOfCells, 0.0);
-  for (size_t i = 0; i < particlesPerCell.size(); i++) {
-    densityPerCell[i] =
-        (allVolumes[i] == 0) ? 0 : (particlesPerCell[i] / allVolumes[i]);  // make sure there is no division of zero
-    if (densityPerCell[i] > maxDensity) {
-      maxDensity = densityPerCell[i];
+  std::vector<double> densityPerBin;
+  densityPerBin.reserve(numberOfBins);
+  for (size_t i = 0; i < numberOfBins; i++) {
+    densityPerBin[i] = (double)particlesPerBin[i] / binVolume;
+    if (densityPerBin[i] > maxDensity) {
+      maxDensity = densityPerBin[i];
     }
   }
 
-  if (maxDensity < 0.0)
+  if (maxDensity < 0.0) {
     throw std::runtime_error("maxDensity can never be smaller than 0.0, but is:" + std::to_string(maxDensity));
```

Comment:
```suggestion
    utils::Exception::exception("calculateHomogeneityAndMaxDensity(): maxDensity can never be smaller than 0.0, but is: {}", maxDensity);
```

Reply:
Done in `c542c145e99aab9fe44b88b8be3f22c83929eaea`